### PR TITLE
Anti-Spawn Camp: Immune to Damage while near Foutain

### PR DESCRIPTION
**Pull request by darklordabc**
There is dozens of way to stop fountain camping this is just one approach. This declares a basic damage filter, and when any hero receives damage it checks if they have the fountain rejuvenation buff and if they do it sets the damage to zero. This also changes the default modifier to give player some information. 
image

The buff can linger for a few seconds after leaving fountain, which means that players can fight outside their area with an advantage, and if they teleport to another player, they will be invulnerable for a couple of seconds. Also, we may want to add a filter on the damage to allow some types of damage being applied while heros are in fountain area, like Zeus's ult prehaps.

### CODE REVIEW

**Comment by Haganeko** [comment]
One question: what about hook?

**Comment by DoctorGester** [comment]
It's just a damage filter. What about hook?

**Comment by darklordabc** [comment]
Hook would still work, but from my knowledge Baumi wanted hooks to still be effective. All spell disables are still enabled, just nothing will cause damage. Also, when players get hooked out, they will still be invulnerable for a few seconds, so they might be able to escape back to fountain.

**Comment by Haganeko** [comment]
I don't know... IMO Hook is the worst part of fountain camping. Especially since no one can go into the fountain to kill people - fountain is too strong for that

**Comment by darklordabc** [comment]
Hooking will be drastically reduced in its effectiveness for a couple of reasons. 
* Players will be invulnerable for about 3 or 4 seconds after getting hooked, so after a successful hook and the enemy team unloading damage, all of it will be ignored, and the hooked player will have 4 seconds to get back to the aura. 
* Camping outside the foutain will become a easier battle to win for the defending team, because of previously mentioned linger of god mode, and the fact they simply have to walk back a bit to get invulnerability. This means defenders can more easily attack heroes who are trying to siege them.

**Comment by darklordabc** [comment]
It would be trivial to prevent hook strategy, like making any players that damage a hero with fountain rejuvenation get teleported back to their base but I think that the current change in the PR will be more enough.

**Comment by chrisinajar** [comment]
addon_english conflicts D:

**Comment by darklordabc** [comment]
Conflicts resolved

**Comment by Chronophylos** [comment]
Nope 
> Conflicting files
game/resource/addon_english.txt

**Comment by darklordabc** [comment]
Should hopefully be fixed

**Comment by chrisinajar** [comment]
don't you love how we get to merge in 1 addon_english change at a time? D: D: D:

**Comment by chrisinajar** [comment]
lololololololololololol

File: game/scripts/vscripts/gamemode.lua [hunk 1]
```diff
@@ -76,7 +76,30 @@ require('components/index')
...
+function GameMode:FilterDamage(filterTable)
+  local victim_index = filterTable["entindex_victim_const"]
+  local attacker_index = filterTable["entindex_attacker_const"]
+  local ability_index = filterTable["entindex_inflictor_const"]
+  if not victim_index or not attacker_index then
+      return true
+  end
+  local victim = EntIndexToHScript(victim_index)
+  local attacker = EntIndexToHScript(attacker_index)
+  print("damaged")
```
Comment (chrisinajar) [inline game/scripts/vscripts/gamemode.lua]: maybe don't print on damage?

Comment (darklordabc) [inline game/scripts/vscripts/gamemode.lua]: removed

**Comment by Chronophylos** [comment]
Don't worry about the changelog.
I'm doing it myself. In fact the version on github is not up-to-date.

**Comment by chrisinajar** [comment]
I don't think the solution to spawn camping is to make the camp stronger anyway, it's better to not have massive power spikes that cause heroes to become so strong they can take on your whole team while fountain healing.

Perhaps we can add this in later, but closing for now to clean things up. It should also probably have been a component.

**Pull request closed without merging**